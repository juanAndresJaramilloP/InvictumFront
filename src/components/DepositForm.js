import React, { useState } from 'react';
import NavBar from './NavBar';
import { useNavigate } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';
import { useEffect } from 'react';

function DepositForm() {
  const [form, setForm] = useState({
    amount: '',
    cardHolderName: '',
    cardNumber: '',
    csv: ''
  });
  const [errors, setErrors] = useState({});
  const navigate = useNavigate();

  const handleInputChange = (e) => {
    const { name, value } = e.target;
    setForm({ ...form, [name]: value });
  };

  const validateForm = () => {
    let tempErrors = {};

    tempErrors.amount = form.amount ? '' : <FormattedMessage id="depositar.error.amount" defaultMessage="Amount is required and it must be a number" />;

    tempErrors.cardHolderName = form.cardHolderName ? '' : <FormattedMessage id="depositar.error.cardholder" defaultMessage="Cardholder name is required" />;


    tempErrors.cardNumber = /^\d{16}$/.test(form.cardNumber) ? '' : <FormattedMessage id="depositar.error.cardnumber" defaultMessage="Card number is invalid. It should have 16 digits" />;

    tempErrors.csv = form.csv && /^\d{3}$/.test(form.csv.toString()) ? '' : <FormattedMessage id="depositar.error.csv" defaultMessage="CSV is invalid and must be a 3 digit number" />;

    setErrors(tempErrors);

    return Object.values(tempErrors).every(x => x === '');
  };


  const handleSubmit = (e) => {
    e.preventDefault();
    if (validateForm()) {
      console.log('Form is valid');
      navigate('/confirmacionDeposito/'+form.amount);
    } else {
      console.log('Form is invalid');
    }
  };
  const handleReturn = () => {
    navigate('/');
  }


  return (
    <div className="">
      <NavBar />
      <button
        onClick={handleReturn}
        className="flex items-center justify-center bg-blue-500 hover:bg-blue-400 text-black font-bold py-2 px-4 m-10 color-gris-fondo rounded focus:outline-none focus:shadow-outline"
      >
        <svg className="w-4 h-4 mr-2 fill-current" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20">
          <path d="M10 15l-5-5 5-5v10zM5 10h10V9H5v1z" />
        </svg>
        <FormattedMessage id="formulario.volver" defaultMessage="Return" />
      </button>


      <div className='flex flex-col items-center justify-center'>
        <div className="w-full max-w-md">
          <div className="flex justify-between items-center mt-4 px-4 w-full"></div>


          <h2 className="text-3xl font-semibold text-center text-gray-700 font-inter -mt-20"><FormattedMessage id="navbar.depositar" defaultMessage="Deposit funds" /></h2>

          <div style={{ width: '72px' }}>
          </div>


          <form className="mt-6" onSubmit={handleSubmit}>
            <div className="mb-4">

              <label htmlFor="amount" className="block text-gray-900 text-sm font-medium mt-20">
                <FormattedMessage id="depositar.amount" defaultMessage="Insert deposit amount in USD" />
              </label>

              <input
                type="number"
                name="amount"

                value={form.amount}
                onChange={handleInputChange}
                className={`w-full px-4 py-2 text-gray-900 bg-transparent border-b-2 ${errors.amount ? 'border-red-500' : 'border-gray-500'
                  } focus:outline-none focus:border-blue-500  mb-3 placeholder-black`}
              />
              {errors.amount && <p className="text-red-500 text-xs mt-1">{errors.amount}</p>}
            </div>
            <div className="mb-4">
              <label htmlFor="amount" className="block text-gray-900  text-sm font-medium mt-11 ">
                <FormattedMessage id="depositar.Cardholder's name" defaultMessage="Cardholder's name" />
              </label>

              <input
                type="text"
                name="cardHolderName"

                value={form.cardHolderName}
                onChange={handleInputChange}
                className={`w-full px-4 py-2 text-gray-900 bg-transparent border-b-2 ${errors.cardHolderName ? 'border-red-500' : 'border-gray-500'
                  } focus:outline-none focus:border-blue-500 mb-3 placeholder-black`}
              />
              {errors.cardHolderName && <p className="text-red-500 text-xs mt-1">{errors.cardHolderName}</p>}
            </div>
            <div className="mb-4">
              <label htmlFor="amount" className="block text-gray-900  text-sm font-medium mt-11 ">
                <FormattedMessage id="depositar.Card number" defaultMessage="Card number" />
              </label>
              <input
                type="text"
                name="cardNumber"
                placeholder=""
                value={form.cardNumber}
                onChange={handleInputChange}
                className={`w-full px-4 py-2 text-gray-900 bg-transparent border-b-2 ${errors.cardNumber ? 'border-red-500' : 'border-gray-500'
                  } focus:outline-none focus:border-blue-500 mb-3 placeholder-black`}
              />
              {errors.cardNumber && <p className="text-red-500 text-xs mt-1">{errors.cardNumber}</p>}
            </div>
            <div className="mb-6">
              <label htmlFor="amount" className="block text-gray-900  text-sm font-medium mt-11 ">
                <FormattedMessage id="depositar.CSV" defaultMessage="CSV" />
              </label>

              <input
                type="number"
                name="csv"

                value={form.csv}
                onChange={handleInputChange}
                className={`w-full px-4 py-2 text-gray-900 bg-transparent border-b-2 ${errors.csv ? 'border-red-500' : 'border-gray-500'
                  } focus:outline-none focus:border-blue-500 mb-3 placeholder-black`}
              />
              {errors.csv && <p className="text-red-500 text-xs mt-1">{errors.csv}</p>}
            </div>
            <div className="mb-6">
              <button
                type="submit"
                className="w-full px-4 my-20 py-5 text-xl text-white bg-blue-500 rounded-3xl hover:bg-blue-600 focus:outline-none color-pagar"
                href="/confirmacionDeposito"
              >
                <FormattedMessage id="depositar.pagar" defaultMessage="Pay" />
              </button>

            </div>
          </form>
        </div>
      </div>

    </div>
  );
}

export default DepositForm;
