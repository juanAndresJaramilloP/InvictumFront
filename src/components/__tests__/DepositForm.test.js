import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { IntlProvider } from 'react-intl';
import DepositForm from "../DepositForm";
import localeEsMessages from "../../locales/es";
import { BrowserRouter } from 'react-router-dom';


jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
      state: { email: 'mrosencwaig0@unesco.org', password: "aA6AM", name: "Malchy Rosencwaig", role : true}
  })
}));

const renderWithRouterAndIntl = (component, locale = 'es', messages = localeEsMessages) => {
  return {
    ...render(
      <BrowserRouter>
        <IntlProvider locale={locale} messages={messages}>
          {component}
        </IntlProvider>
      </BrowserRouter>
    )
  };
};

describe('DepositForm', () => {

  it('updates input field values amount', () => {
    renderWithRouterAndIntl(<DepositForm />);
    const amountInput = screen.getByRole('spinbutton', { name: /Ingrese la cantidad a depositar en USD/ });
    fireEvent.change(amountInput, { target: { value: '123' } });
    expect(amountInput).toHaveValue(123);
  });
  it('updates input field values for card holder name', () => {
    renderWithRouterAndIntl(<DepositForm />);
    const cardHolderNameInput = screen.getByRole('textbox', { name: /Nombre del titular de la tarjeta/ });
    fireEvent.change(cardHolderNameInput, { target: { value: 'Juan Pérez' } }); 
    expect(cardHolderNameInput).toHaveValue('Juan Pérez');
  });
  it('updates input field values for card number', () => {
    renderWithRouterAndIntl(<DepositForm />);
    const cardNumberInput = screen.getByRole('textbox', { name: /Numero de tarjeta/ });
    fireEvent.change(cardNumberInput, { target: { value: '1234567890123456' } });
    expect(cardNumberInput).toHaveValue('1234567890123456');
  });
  
  it('updates input field values for CSV', () => {
    renderWithRouterAndIntl(<DepositForm />);
    const csvInput = screen.getByRole('spinbutton', { name: /CSV/ });
    fireEvent.change(csvInput, { target: { value: '123' } });
    expect(csvInput).toHaveValue(123);
  });
  
  

  it('validates fields correctly when empty', () => {
    renderWithRouterAndIntl(<DepositForm />);
    const submitButton = screen.getByRole('button', { name: /Pagar/ });

    fireEvent.click(submitButton);


    expect(screen.queryByText(/La cantidad es requerida y debe ser un numero./i)).toBeInTheDocument();
    expect(screen.queryByText(/El nombre del titular de la tarjeta es requerido./i)).toBeInTheDocument();
    expect(screen.queryByText(/El numero de tarjeta es requerido. Debe ser un numero de 16 digitos./i)).toBeInTheDocument();
    expect(screen.queryByText(/El CSV es requerido. Debe ser un numero de 3 digitos./i)).toBeInTheDocument();
    
  });

});
