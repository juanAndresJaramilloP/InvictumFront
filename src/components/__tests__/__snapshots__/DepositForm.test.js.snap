// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DepositForm renders correctly with localized texts 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="navbar bg-[#030A1C] text-white"
    >
      <div
        class="navbar-start ml-8"
      >
        <div
          class="dropdown"
        >
          <div
            class="btn btn-ghost md:hidden"
            role="button"
            tabindex="0"
          >
            <svg
              class="h-5 w-5"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6h16M4 12h8m-8 6h16"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </div>
          <ul
            class="menu menu-sm dropdown-content mt-3 z-[1] p-2 shadow bg-base-100 rounded-box w-52 text-black"
            tabindex="0"
          >
            <li>
              <a>
                Item 1
              </a>
            </li>
            <li>
              <a>
                Parent
              </a>
              <ul
                class="p-2 text-black"
              >
                <li>
                  <a>
                    Submenu 1
                  </a>
                </li>
                <li>
                  <a>
                    Submenu 2
                  </a>
                </li>
              </ul>
            </li>
            <li>
              <a>
                Item 3
              </a>
            </li>
          </ul>
        </div>
        <img
          alt="Logo"
          class="h-20"
          src="Invictum.svg"
        />
        <a
          class="btn btn-ghost p-0 text-xl"
        >
          NVICTUM
        </a>
      </div>
      <div
        class="navbar-center hidden md:flex"
      >
        <ul
          class="menu menu-horizontal px-1"
        >
          <li>
            <a
              href="/Aprendizaje"
            >
              Educacion
            </a>
          </li>
          <li>
            <details>
              <summary>
                Transacciones
              </summary>
              <ul
                class="p-2 text-black w-32"
              >
                <li>
                  <a>
                    Depositar fondos
                  </a>
                </li>
                <li>
                  <a>
                    Retirar fondos
                  </a>
                </li>
              </ul>
            </details>
          </li>
          <li>
            <a>
              Reportes
            </a>
          </li>
        </ul>
      </div>
      <div
        class="navbar-end flex mr-8"
      >
        <p
          class="text-xl mr-4"
        />
        <details
          class="dropdown dropdown-end"
        >
          <summary>
            <img
              alt="Profile Image"
              class="h-20 bg-black rounded-full p-3 mr-4"
              src="profileImage.png"
            />
          </summary>
          <ul
            class="p-2 shadow menu dropdown-content z-[1] bg-base-100 rounded-box w-52"
          >
            <li>
              <a
                class="text-black"
              >
                Configuración de administrador
              </a>
            </li>
            <li>
              <a
                class="text-black"
              >
                Reestablecer Contraseña
              </a>
            </li>
          </ul>
        </details>
      </div>
    </div>
    <button
      class="flex items-center justify-center bg-blue-500 hover:bg-blue-400 text-black font-bold py-2 px-4 m-10 color-gris-fondo rounded focus:outline-none focus:shadow-outline"
    >
      <svg
        class="w-4 h-4 mr-2 fill-current"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 15l-5-5 5-5v10zM5 10h10V9H5v1z"
        />
      </svg>
      Volver
    </button>
    <div
      class="flex flex-col items-center justify-center"
    >
      <div
        class="w-full max-w-md"
      >
        <div
          class="flex justify-between items-center mt-4 px-4 w-full"
        />
        <h2
          class="text-3xl font-semibold text-center text-gray-700 font-inter -mt-20"
        >
          Depositar fondos
        </h2>
        <div
          style="width: 72px;"
        />
        <form
          class="mt-6"
        >
          <div
            class="mb-4"
          >
            <label
              class="block text-gray-900 text-sm font-medium mt-20"
              for="amount"
            >
              Ingrese la cantidad a depositar en USD
            </label>
            <input
              class="w-full px-4 py-2 text-gray-900 bg-transparent border-b-2 border-gray-500 focus:outline-none focus:border-blue-500  mb-3 placeholder-black"
              id="amount"
              name="amount"
              type="number"
              value=""
            />
          </div>
          <div
            class="mb-4"
          >
            <label
              class="block text-gray-900  text-sm font-medium mt-11 "
              for="amount"
            >
              Nombre del titular de la tarjeta
            </label>
            <input
              class="w-full px-4 py-2 text-gray-900 bg-transparent border-b-2 border-gray-500 focus:outline-none focus:border-blue-500 mb-3 placeholder-black"
              name="cardHolderName"
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-4"
          >
            <label
              class="block text-gray-900  text-sm font-medium mt-11 "
              for="amount"
            >
              Numero de tarjeta
            </label>
            <input
              class="w-full px-4 py-2 text-gray-900 bg-transparent border-b-2 border-gray-500 focus:outline-none focus:border-blue-500 mb-3 placeholder-black"
              name="cardNumber"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-6"
          >
            <label
              class="block text-gray-900  text-sm font-medium mt-11 "
              for="amount"
            >
              CSV
            </label>
            <input
              class="w-full px-4 py-2 text-gray-900 bg-transparent border-b-2 border-gray-500 focus:outline-none focus:border-blue-500 mb-3 placeholder-black"
              name="csv"
              type="number"
              value=""
            />
          </div>
          <div
            class="mb-6"
          >
            <button
              class="w-full px-4 my-20 py-5 text-xl text-white bg-blue-500 rounded-3xl hover:bg-blue-600 focus:outline-none color-pagar"
              href="/confirmacionDeposito"
              type="submit"
            >
              Pagar
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
